Measured boot update for Linux OS Image

lsubash opened the pull request:
Unified interface is added to extend OS Image digest.
For Container OS Image this interface is added as
callback functionality. IAS image utilizes same
functionality to extend TPM digest.

Signed-off-by: Subash Lakkimsetti <subash.lakkimsetti@intel.com>

mauricema commented on BootloaderCommonPkg/Library/ContainerLib/ContainerLib.c:
```diff
@@ -397,6 +399,19 @@ AutheticateContainerInternal (
         Status   = AuthenticateComponent (DataBuf, DataLen, CompEntry->AuthType,
                                           AuthData, CompEntry->HashData, 0);
+
+        if(Status == EFI_SUCCESS) {
```
space after "if"

mauricema replied:
should use (!EFI_ERROR(Status))

mauricema commented on BootloaderCommonPkg/Library/ContainerLib/ContainerLib.c:
```diff
@@ -397,6 +399,19 @@ AutheticateContainerInternal (
+          CbInfo.CompLen          = DataLen;
+          CbInfo.HashAlg          = GetHashAlg(CompEntry->AuthType);
+          CbInfo.HashData         = CompEntry->HashData;
+          if (ContainerCallback != NULL) {
```
move condition to be before #406

mauricema commented on PayloadPkg/OsLoader/OsLoader.c:
```diff
@@ -10,6 +10,56 @@
+  //Convert Measured boot Hash Mask to TPMI_ALG_HASH (TPM ALG ID)
+  MbTmpAlgHash = (TPMI_ALG_HASH) GetTpmHashAlg(PcdGet32(PcdMeasuredBootHashMask));
+
+  LoaderPlatformInfo = (LOADER_PLATFORM_INFO *)GetLoaderPlatformInfoPtr();
```
Create abstraction layer so that it does not need LAODER_PLATFORM_INFO.   Then the code can be common between Stage and Payload.

lsubash replied:
BootloaderCore has MEASURED_BOOT_ENABLED() support and it currently getting used in stage components. 
Check for measured boot using LAODER_PLATFORM_INFO would be limited to the Osloader only.

myguitar replied:
Please use GetFeatureCfg() instead of LOADER_PLATFORM_INFO.

lsubash replied:
Thanks Aiden. I will check.

lsubash commented:
Updated patch with Maurice comments except last one with abstraction of LAODER_PLATFORM_INFO in OSLoader

myguitar commented on BootloaderCommonPkg/Include/Library/ContainerLib.h:
```diff
@@ -211,7 +213,8 @@ GetNextAvailableComponent (
 RegisterContainer (
-  IN  UINT32   ContainerBase
+  IN  UINT32               ContainerBase,
+  IN  CONTAINER_CALLBACK   ContainerCallback
```
Please update parameter description as well.

myguitar commented on BootloaderCommonPkg/Include/Library/IasImageLib.h:
```diff
@@ -93,9 +101,9 @@ hash of the IAS image to the caller.
-  OUT UINT8 *ImageHash
+  IN  VOID               *Addr,
+  IN  UINT32              Size,
+  OUT IAS_IMAGE_INFO     *IasImageInfo
```
Please update parameter description as well.

myguitar commented on BootloaderCommonPkg/Library/ContainerLib/ContainerLib.c:
```diff
@@ -341,7 +341,8 @@ GetContainerKeyUsageBySig (
 AutheticateContainerInternal (
-  IN  CONTAINER_HDR      *ContainerHeader
+  IN  CONTAINER_HDR      *ContainerHeader,
+  IN  CONTAINER_CALLBACK  ContainerCallback
```
Please update parameter description as well.

myguitar commented on BootloaderCommonPkg/Library/IasImageLib/IasImage.c:
```diff
@@ -30,9 +30,9 @@ hash of the IAS image to the caller.
-  OUT UINT8    *ImageHash
+  IN  VOID               *ImageAddr,
+  IN  UINT32              Size,
+  OUT IAS_IMAGE_INFO     *IasImageInfo
```
Please update parameter description as well.

lsubash commented:
Updated patch as per Aiden's feedback.

lsubash commented:
Updated PR with following,
- Moved ExtendStageHash() to bootloader common and same functionality is used in Stage1B, Stage2 and OsLoader while extending TPM hash.
- Updated Osloader to use LoadComponentCallback call back instead of defining a new one.
- For Linux Os in container format, Container header and Container hash gets extended to PCR(8).

mauricema commented on BootloaderCommonPkg/Library/BootloaderCommonLib/BootloaderCommonLib.c:
```diff
@@ -783,3 +785,62 @@ GetHashToExtend (
 
   return Status;
 }
+
```
I think this should be in some of the security related library.  Not common lib here.

lsubash replied:
Done. Moved this API to TpmLib.
